### PR TITLE
RuneLite

### PR DESCRIPTION
**Pull request by rushsteve1**
RuneLite OSRS client.
I was having some trouble setting up some persistent files. RuneLite automatically puts 2 files and a folder in the user's home folder which I wanted to avoid doing, but I cannot figure out how to make them persistent in the Flatpak. Some help on this would be greatly appreciated.

### CODE REVIEW

**Comment by TingPing** [comment]
> RuneLite automatically puts 2 files and a folder in the user's home folder which I wanted to avoid doing, but I cannot figure out how to make them persistent in the Flatpak. Some help on this would be greatly appreciated.

You have to resort to hacks if it makes single files in `$HOME` like a wrapper script that sets `$HOME` to a private directory in `$XDG_CONFIG_HOME` assuming it respects that env var.

File: net.runelite.RuneLite.desktop [hunk 1]
```diff
@@ -0,0 +1,9 @@
+[Desktop Entry]
```
Comment (TingPing) [inline net.runelite.RuneLite.desktop]: You have data files but never install them.

File: net.runelite.RuneLite.json [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+            "build-commands": [
+                "install -D runelite /app/bin/runelite",
+                "install -D RuneLite.jar /app/share/RuneLite.jar"
+            ],
+            "sources": [
+                {
+                    "type": "script",
+                    "dest-filename": "runelite",
+                    "commands": [
+                        "ls -R /app | grep jagex",
```
Comment (AsciiWolf) [inline net.runelite.RuneLite.json]: What is this good for? Seems to be just a check if the directory exist.

File: net.runelite.RuneLite.json [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+        "org.freedesktop.Sdk.Extension.openjdk10" 
+    ],
+    "finish-args": [
+        "--share=ipc",
+        "--share=network",
+        "--socket=x11",
+        "--socket=wayland",
+        "--socket=pulseaudio",
+        "--env=JAVA_HOME=/app/jre",
+        "--persist=~"
```
Comment (AsciiWolf) [inline net.runelite.RuneLite.json]: I am not sure if this is actually used. The RuneLite updater seems to be downloading all the game files every time.

File: net.runelite.RuneLite.desktop (hunk 1 above)
Comment (AsciiWolf) [reply net.runelite.RuneLite.desktop]: Yep, same thing with the appdata file.

File: net.runelite.RuneLite.json (hunk 3 above)
Comment (AsciiWolf) [reply net.runelite.RuneLite.json]: (You can make multiple `--persist` options, for example `--persist=.foo, --persist=.bar` will make the ~/.foo and ~/.bar files/directories persistent.)

**Comment by AsciiWolf** [comment]
<pre>diff --git a/net.runelite.RuneLite.desktop b/net.runelite.RuneLite.desktop
index 89d5cc5..511562c 100644
--- a/net.runelite.RuneLite.desktop
+++ b/net.runelite.RuneLite.desktop
@@ -6,4 +6,4 @@ Comment=RuneLite is a free, open-source and super fast client for Old School Run
 Keywords=games;runescape;osrs;
 Exec=runelite
 Icon=net.runelite.RuneLite
-Categories=Game;MMO;
+Categories=Game;RolePlaying;
diff --git a/net.runelite.RuneLite.json b/net.runelite.RuneLite.json
index 089da41..bb48bd7 100644
--- a/net.runelite.RuneLite.json
+++ b/net.runelite.RuneLite.json
@@ -30,14 +30,16 @@
             "buildsystem": "simple",
             "build-commands": [
                 "install -D runelite /app/bin/runelite",
-                "install -D RuneLite.jar /app/share/RuneLite.jar"
+                "install -D RuneLite.jar /app/share/RuneLite.jar",
+                "install -Dm644 net.runelite.RuneLite.png /app/share/icons/hicolor/128x128/apps/net.runelite.RuneLite.png",
+                "install -Dm644 net.runelite.RuneLite.desktop /app/share/applications/net.runelite.RuneLite.desktop",
+                "install -Dm644 net.runelite.RuneLite.appdata.xml /app/share/appdata/net.runelite.RuneLite.appdata.xml"
             ],
             "sources": [
                 {
                     "type": "script",
                     "dest-filename": "runelite",
                     "commands": [
-                        "ls -R /app | grep jagex",
                         "exec $JAVA_HOME/bin/java -jar /app/share/RuneLite.jar"
                     ]
                 },
@@ -45,6 +47,18 @@
                     "type": "file",
                     "url": "https://github.com/runelite/launcher/releases/download/1.6.0/RuneLite.jar",
                     "sha256": "fdcfdd51b033815c4f45025c315175b8a20cc080b65a8b76aee5519b39e85d60"
+                },
+                {
+                    "type": "file",
+                    "path": "net.runelite.RuneLite.png"
+                },
+                {
+                    "type": "file",
+                    "path": "net.runelite.RuneLite.desktop"
+                },
+                {
+                    "type": "file",
+                    "path": "net.runelite.RuneLite.appdata.xml"
                 }
             ]
         }</pre>

Here's a fix to make the desktop/appdata files working (and also to remove the unnecessary check).

Persistent RuneLite files are still an issue; if they have always the same name, you can use `--persist`, but if their name is changing, you have to use some hack like TingPing suggested.

File: net.runelite.RuneLite.appdata.xml [hunk 4]
```diff
@@ -0,0 +1,23 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (AsciiWolf) [inline net.runelite.RuneLite.appdata.xml]: Also, please add [OARS](https://hughsie.github.io/oars/) metadata to the appdata file.

**Comment by rushsteve1** [comment]
@AsciiWolf Thank you for your patch! And the files do not change name as far as I can tell, but when I tried using `--persist` on them it generated folders of the same name instead and persisted those.

**Comment by AsciiWolf** [comment]
@rushsteve1 Ah, weird, I thought that `--persist` supports files as well. :-/

File: net.runelite.RuneLite.desktop [hunk 5]
```diff
@@ -0,0 +1,9 @@
...
+Type=Application
+Terminal=false
+Name=RuneLite
+Comment=RuneLite is a free, open-source and super fast client for Old School RuneScape.
+Keywords=games;runescape;osrs;
+Exec=runelite
+Icon=net.runelite.RuneLite
+Categories=Game;MMO;
```
Comment (AsciiWolf) [inline net.runelite.RuneLite.desktop]: Should be `Game;RolePlaying;`. It is fixed in my patch.

File: net.runelite.RuneLite.json (hunk 2 above)
Comment (rushsteve1) [reply net.runelite.RuneLite.json]: Oh. I forgot to remove that.

**Comment by rushsteve1** [comment]
Ok last commit merged @AsciiWolf's patch and added OARS data. I also tried setting a `HOME` in the launcher script to a folder and then persisting that folder but it didn't seem to work so any more suggestions would be greatly appreciated.
Because to be honest those files cluttering my home folder were what prompted me to do this :upside_down_face:

File: net.runelite.RuneLite.desktop (hunk 1 above)
Comment (rushsteve1) [reply net.runelite.RuneLite.desktop]: This was fixed.

File: net.runelite.RuneLite.appdata.xml [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+
+    <screenshots>
+        <screenshot type="default">https://i.imgur.com/oaQEtYw.png</screenshot>
+    </screenshots>
+
+    <releases>
+        <release version="1.6.0" date="2018-7-7"></release>
+    </releases>
+    <update_contact>rushsteve1_at_teknik.io</update_contact>
+
```
Comment (AsciiWolf) [inline net.runelite.RuneLite.appdata.xml]: Unnecessary empty line. :-)

**Comment by AsciiWolf** [comment]
@rushsteve1: It's a pity that the persist workaround didn't work. :-/ Maybe @alexlarsson will know more about this?

**Comment by rushsteve1** [comment]
I figured out how to persist the downloaded data and the jagexcache which are both folders. The downloaded data was the big thing.
To be honest I have no idea what the two files I cannot persist actually do and they don't seem to affect anything when removed.

**Comment by nedrichards** [comment]
bot, build net.runelite.RuneLite

**Comment by AsciiWolf** [comment]
I can confirm that the issue with persistent files is now fixed. Game is fully playable, however, there are still some minor issues:
- Button to recover account password does not work.
- UI is a little glitchy (does not prevent the actual gameplay).
- Tray icon seems to be empty.

**Comment by rushsteve1** [comment]
In response to @AsciiWolf's concerns:
- The button to recover account password is either an issue opening the browser or with the game itself. Can you create a new account from the game?
- Do you mean the right sidebar UI? If so then this is a known RuneLite bug and happens on all platforms.
- I might need to rename the icon for it to work, I'll check on that.

**Comment by AsciiWolf** [comment]
@rushsteve1:
- I didn't try to create a new account from the game and I don't have any e-mail address available that I can use to test it.
- Yes, the right sidebar UI.
- I found out that the icon is displayed, but placed on the left side of my desktop. Weird, maybe some bug in the TopIcons Plus GNOME Shell extension that I use.

**Comment by rushsteve1** [comment]
The Password recover issue is the same as when you click "Create Account" it wont open the browser. I have no idea what would cause that.

**Comment by rushsteve1** [comment]
I assume that the driver issues are attempts at loading some form of hardware acceleration. While that would be nice to have I know nothing about Flatpak and drivers, and to be blunt this is a browser game from 2007. The performance requirements are minimal.

The `javax.net.ssl.SSLHandshakeException` are probably just dropped packets or something, I am not terribly worried about it.

**Comment by AsciiWolf** [comment]
@rushsteve1 I don't see any problem in the json file. But I found out that RuneLite devs itself recommend disabling 3D acceleration in some cases and RuneLite probably works fine without 3D acceleration.

**Comment by AsciiWolf** [comment]
Anyway, besides the browser issue and the glitched UI that is an upstream issue, it seems to work fine so this PR can probably be merged. ;-)

**Comment by rushsteve1** [comment]
@TingPing @AsciiWolf Is there anything else that needs to be done before this is merged?

**Comment by nedrichards** [comment]
Repository has been created: https://github.com/flathub/net.runelite.RuneLite

**Pull request closed without merging**